### PR TITLE
Speed up kw kernel build

### PR DESCRIPTION
**Pull request by isinyaaa**
This PR addresses the main points of issue #158, but also some other features that were missing, like support for warnings and the LLVM toolchain.

### CODE REVIEW

File: src/build.sh [hunk 1]
```diff
@@ -89,12 +89,18 @@ function kernel_build()
   fi
 
   if [ -x "$(command -v nproc)" ]; then
-    parallel_cores=$(nproc --all)
```
Comment (rodrigosiqueira) [inline src/build.sh:90]: Since you are working on this part of the code, we can replace `-x "$(command ...` with `command_exists`.

File: src/build.sh [hunk 2]
```diff
@@ -89,12 +89,18 @@ function kernel_build()
...
+    parallel_cores="-j$(nproc --all)"
   else
-    parallel_cores=$(grep -c ^processor /proc/cpuinfo)
+    parallel_cores="-j$(grep -c ^processor /proc/cpuinfo)"
   fi
 
-  command="make -j$parallel_cores ARCH=$arch $cross_compile"
+  if [[ -n "${options_values['OPTIMIZE']}" ]]; then
+    optimizations="CC=\"ccache gcc -fdiagnostics-color\" $parallel_cores"
```
Comment (rodrigosiqueira) [inline src/build.sh]: I would prefer that we handle `ccache` as a separate option. We can have an entry in the `etc/kworkflow_template.config` to enable it (e.g., `enable_ccache=no`). We can also have a new option under the `kw b` named `--ccache`. Finally, you need to update files in the folder `documentation/dependencies/` to make ccache explicit kw dependencies.

File: src/build.sh [hunk 3]
```diff
@@ -89,12 +89,18 @@ function kernel_build()
...
+  else
+    optimizations="$parallel_cores"
```
Comment (rodrigosiqueira) [inline src/build.sh:116]: I think we should also allow users to say how many percent of their CPU they want to utilize, for example, something like `kw build --cpu-utilization "50%"`. I want full-speed compilation most of the time, but sometimes I want to keep some CPU resources for other tasks.

File: src/build.sh [hunk 4]
```diff
@@ -130,7 +136,7 @@ function parse_build_options()
   options_values['ARCH']="${configurations[arch]:-'x86_64'}"
   options_values['MENU_CONFIG']=''
   options_values['CROSS_COMPILE']="${configurations[cross_compile]}"
-  options_values['PARALLEL_CORES']=1
+  options_values['OPTIMIZE']="${configurations[optimize]}"
```
Comment (rodrigosiqueira) [inline src/build.sh:155]: The associative array `configurations` is associated with `kworkflow.config` file, in this sense, you need to add `optimize` in the `etc/kworkflow_template.config`. Also, we can use this option to store user-specific flags.

File: src/build.sh [hunk 5]
```diff
@@ -47,6 +47,7 @@ function kernel_build()
   local arch
   local menu_config
   local doc_type
+  local llvm
```
Comment (rodrigosiqueira) [inline src/build.sh:50]: This is a very nice commit :)
Let's add these compiler options in the `kworkflow.config` file.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
First of all, thanks a lot for your PR; I really like your PR, and it gave me some ideas.

How about making the build option more flexible? Let's create a new config file dedicated to the compilation setup, something like this:

.kw/build.config

In this config file, we can set things like:

* Optimization flags
* ccache
* Warning level
* Target compiler

This code refactor can make your work with the env option a little bit easier.

File: src/build.sh (hunk 3 above)
Comment (isinyaaa) [reply src/build.sh:116]: But do you think a percentage would work alright? How do you propose we make those calculations?

**Comment by rodrigosiqueira** [comment]
Hi @isinyaaa any progress here?

**Comment by isinyaaa** [comment]
Hi, @rodrigosiqueira, I should be getting back to this PR in the next few weeks. I'm sorry for the slow pacing!

File: etc/build_template.config [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+# Set default warning level for compilation tasks
+#warning_level=1
+
+# Use this flag as a way to indicate to kw the kernel image name. If this
+# option is not specified, kw will try to infer the correct file inside
+# arch/*/boot/.
+kernel_img_name=bzImage
+
+# Specify cross-compile name
```
Comment (rodrigosiqueira) [inline etc/build_template.config:21]: Could you replace "name" with "toolchain"?

File: src/config_manager.sh [hunk 7]
```diff
@@ -287,11 +287,11 @@ function get_config_from_defconfig()
...
   fi
 
-  cross_compile=${configurations[cross_compile]}
-  arch=${configurations[arch]}
+  cross_compile=${build_config[cross_compile]}
+  arch=${build_config[arch]}
   # TODO: Right now, the below variable has no meaning and will be empty. Let's
   # expand the config file to enable developers to define ASIC targets.
-  asic=${configurations[asic]}
+  asic=${build_config[asic]}
```
Comment (rodrigosiqueira) [inline src/config_manager.sh:294]: tbh, I don't remember why I added it, and this is an excellent example of why we should open issues haha. Let's drop these lines.

File: src/kw_config_loader.sh [hunk 8]
```diff
@@ -159,11 +183,11 @@ function load_configuration()
...
   local -a config_dirs
   local config_dirs_size
-  local IFS=:
-  read -ra config_dirs <<< "${XDG_CONFIG_DIRS:-"/etc/xdg"}"
-  unset IFS
+
+  IFS=: read -ra config_dirs <<< "${XDG_CONFIG_DIRS:-"/etc/xdg"}"
 
   parse_configuration "$KW_ETC_DIR/$CONFIG_FILENAME"
+  parse_configuration "$KW_ETC_DIR/$BUILD_CONFIG_FILENAME" build_config
```
Comment (rodrigosiqueira) [inline src/kw_config_loader.sh:190]: What if we move the `build.config` load to `src/build.sh`? In the current implementation, we are loading all config every time that we include this file (`kw_cionfig_loader`). Maybe we can make the load option per feature and save some CPU cycles.

File: documentation/dependencies/debian.dependencies [hunk 9]
```diff
@@ -24,3 +24,4 @@ bc
 git-email
 sqlite3
 pv
+ccache
```
Comment (rodrigosiqueira) [inline documentation/dependencies/debian.dependencies:27]: We now have `fedora.dependencies` as well.

File: src/build.sh [hunk 10]
```diff
@@ -151,6 +162,20 @@ function parse_build_options()
         options_values['MENU_CONFIG']="${build_config[menu_config]:-nconfig}"
         shift
         ;;
+      --cpu-util | --cpu-scale | -c)
```
Comment (rodrigosiqueira) [inline src/build.sh:183]: Why do we want two names here?
Also, you forgot to update the help function and the man page.

File: src/build.sh [hunk 11]
```diff
@@ -151,6 +162,20 @@ function parse_build_options()
...
+        [[ ! "$2" =~ [0-9]+ ]] && {
+          options_values['ERROR']="$2"
+          return 22
+        }
```
Comment (rodrigosiqueira) [inline src/build.sh:187]: Let's use the standard if here. Also, add the `# ENVAL` comment after 22.

File: src/build.sh [hunk 12]
```diff
@@ -151,6 +162,20 @@ function parse_build_options()
...
+        [ "$2" -gt 100 ] &&
```
Comment (rodrigosiqueira) [inline src/build.sh:188]: Same as the previous comment, and why not use `[[ .. ]]`?

File: tests/build_test.sh [hunk 13]
```diff
@@ -148,8 +148,14 @@ function test_kernel_build_x86()
...
   )
-
   compare_command_sequence '' "$LINENO" 'expected_cmd' "$output"
+
+  output=$(kernel_build 'TEST_MODE' --ccache | tail -n +1 | head -2) # Remove statistics output
+  declare -a expected_cmd=(
+    "make -j ARCH=x86_64  olddefconfig --silent"
+    "make CC=\"ccache gcc -fdiagnostics-color\" -j$PARALLEL_CORES ARCH=x86_64"
+  )
+  compare_command_sequence '' "($LINENO)" "expected_cmd" "$output"
```
Comment (rodrigosiqueira) [inline tests/build_test.sh:158]: Create a separate test function for that.\ and add a test for `--cpu-scale` scenario.

File: src/build.sh [hunk 14]
```diff
@@ -70,11 +70,11 @@ function kernel_build()
   fi
 
   if [[ -n "$cross_compile" ]]; then
-    cross_compile="CROSS_COMPILE=$cross_compile"
+    cross_compile=" CROSS_COMPILE=$cross_compile "
   fi
 
   if [[ -n "$menu_config" ]]; then
-    command="make -j ARCH=$arch $cross_compile $menu_config"
+    command="make -j ARCH=$arch$cross_compile$menu_config"
```
Comment (rodrigosiqueira) [inline src/build.sh:80]: Since you are concatenating variables, let's use: `${arch}${cross_compile}${menu_config}`.
https://kworkflow.org/content/codingstyle.html#string-concatenation
(same for this commit)

File: src/build.sh [hunk 15]
```diff
@@ -99,12 +110,17 @@ function kernel_build()
   fi
 
   if [[ -n "${options_values['CCACHE']}" ]]; then
-    optimizations="CC=\"ccache gcc -fdiagnostics-color\" $parallel_cores"
+    if [[ -n "$llvm" ]]; then
+      ccache='CC="ccache clang -fdiagnostics-color"'
+    else
+      ccache='CC="ccache gcc -fdiagnostics-color"'
+    fi
```
Comment (rodrigosiqueira) [inline src/build.sh:117]: Maybe to avoid coding duplication, we could have a' compiler' variable that uses gcc by default (and you can set `clang` inside the `$llvm` condition). After that, you just need to replace the default compiler and build the `ccache` based on that (e.g., `ccache $compiler -fdiagnostics-color"`).

File: src/build.sh [hunk 16]
```diff
@@ -176,10 +194,28 @@ function parse_build_options()
...
+        ;;
+      --optimize | -o)
+        options_values['OPTIMIZE']=1
+        shift
+        ;;
       --doc | -d)
         options_values['DOC_TYPE']="${build_config[doc_type]:-htmldocs}"
         shift
         ;;
+      --warnings | -w)
```
Comment (rodrigosiqueira) [inline src/build.sh:209]: Do not forget to update the help and man.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hi @isinyaaa ,
First of all, thanks a lot for this PR! This is a very exciting change, and this change will unblock my work of enabling deploy support to ChromeOS. :)

I added a couple of comments, but could you add one extra commit redirecting all compilation output to a file? Something like `kw b -w 1 --save-log-to <PATH>`.
This could be useful for newcomers looking for new contributions, such as fixing compilation warnings.

File: src/build.sh (hunk 12 above)
Comment (isinyaaa) [reply src/build.sh:188]: I feel that using bash syntax for something as simple is kinda unnecessary (i.e. `[[` is a bashism, and not a simple alternative to `[`)

File: src/build.sh (hunk 15 above)
Comment (isinyaaa) [reply src/build.sh:117]: really nice suggestion! thanks :)

File: src/kw_config_loader.sh (hunk 8 above)
Comment (isinyaaa) [reply src/kw_config_loader.sh]: took me a while to do this, and I think some tests didn't end up as they were supposed to. Maybe if we could merge the test fixing PR before this one it'd help in this review.

File: src/build.sh (hunk 10 above)
Comment (isinyaaa) [reply src/build.sh]: I'm leaving docs for last, sorry about that.

File: tests/build_test.sh [hunk 17]
```diff
@@ -356,7 +385,7 @@ function test_parse_build_options()
   assert_equals_helper 'Default ARCH did not match expectation' "($LINENO)" 'x86_64' "${options_values['ARCH']}"
   assert_equals_helper 'Default MENU_CONFIG did not match expectation' "($LINENO)" '' "${options_values['MENU_CONFIG']}"
   assert_equals_helper 'Default CROSS_COMPILE did not match expectation' "($LINENO)" '' "${options_values['CROSS_COMPILE']}"
-  assert_equals_helper 'Default PARALLEL_CORES did not match expectation' "($LINENO)" '1' "${options_values['PARALLEL_CORES']}"
```
Comment (isinyaaa) [inline tests/build_test.sh]: TODO: change those assertions

File: src/build.sh [hunk 18]
```diff
@@ -86,27 +91,39 @@ function kernel_build()
...
+  fi
+
   if command_exists nproc; then
     parallel_cores="$(nproc --all)"
   else
     parallel_cores="$(grep -c ^processor /proc/cpuinfo)"
   fi
 
-  parallel_cores="-j$((parallel_cores * cpu_scaling_factor / 100))"
+  optimizations="-j$((parallel_cores * cpu_scaling_factor / 100))"
```
Comment (isinyaaa) [inline src/build.sh:112]: TODO: put that in the optimizations commit

File: src/build.sh [hunk 19]
```diff
@@ -86,27 +91,39 @@ function kernel_build()
...
+
+  if [[ -n "${options_values['CCACHE']}" ]]; then
+    compiler='gcc'
+    [[ -n "$llvm" ]] && compiler='clang'
+    optimizations="CC=\"ccache $compiler -fdiagnostics-color\" $optimizations"
+  fi
 
   if [[ -n "$doc_type" ]]; then
-    command="make $parallel_cores $doc_type"
+    command="make $optimizations $doc_type"
```
Comment (isinyaaa) [inline src/build.sh:121]: also this

File: src/build.sh [hunk 20]
```diff
@@ -147,6 +164,9 @@ function parse_build_options()
   options_values['CROSS_COMPILE']="${build_config[cross_compile]}"
   options_values['CCACHE']="${build_config[ccache]}"
   options_values['CPU_SCALING_FACTOR']="${build_config[cpu_scaling_factor]:-100}"
+  options_values['USE_LLVM_TOOLCHAIN']="${build_config[use_llvm_toolchain]}"
+  options_values['SAVE_LOG_TO']="${build_config[save_log_to]}"
```
Comment (isinyaaa) [inline src/build.sh:168]: TODO: add this option to the templates and also test it

File: src/build.sh [hunk 21]
```diff
@@ -147,6 +164,9 @@ function parse_build_options()
   options_values['CROSS_COMPILE']="${build_config[cross_compile]}"
   options_values['CCACHE']="${build_config[ccache]}"
   options_values['CPU_SCALING_FACTOR']="${build_config[cpu_scaling_factor]:-100}"
+  options_values['USE_LLVM_TOOLCHAIN']="${build_config[use_llvm_toolchain]}"
```
Comment (isinyaaa) [inline src/build.sh:167]: TODO: test the LLVM, warnings and option composition.

**Comment by isinyaaa** [comment]
I think this PR is pretty much ready, now, as I've resolved all remaining TODOs. Looking forward to your reviews :)

**Comment by rodrigosiqueira** [comment]
Hi @isinyaaa ,
Could you rebase this branch with the latest version of the unstable branch?

**Comment by isinyaaa** [comment]
@rodrigosiqueira Done ;)

File: tests/build_test.sh [hunk 22]
```diff
@@ -53,8 +64,189 @@ tearDown()
...
+  parse_configuration "$TMP_DIR/build.config" build_config
+  assertConfigurations build_config expected_build_configurations "$LINENO"
+
+  true # Reset return value
+}
+
+function assertConfigurations()
+{
+  declare -n configurations_ref=$1
+  declare -n expected_configurations_ref=$2
```
Comment (rodrigosiqueira) [inline tests/build_test.sh:98]: hmmm... don't we need `""` around `$1` and `$2`?

File: tests/build_test.sh [hunk 23]
```diff
@@ -53,8 +64,189 @@ tearDown()
...
+  local lineno=${3:-LINENO}
```
Comment (rodrigosiqueira) [inline tests/build_test.sh:99]: I think `LINENO` here will not work as expected; basically, it will point to this line instead of where `assertConfigurations` was invoked.

File: tests/build_test.sh [hunk 24]
```diff
@@ -53,8 +64,189 @@ tearDown()
...
+
+  # check if configurations is contained in expected_configurations
+  for k in "${!configurations_ref[@]}"; do
+    if [[ ${expected_configurations_ref[$k]+token} != token ]]; then
```
Comment (rodrigosiqueira) [inline tests/build_test.sh:103]: `+token`? That's new for me. Could you elaborate on this?

File: tests/build_test.sh [hunk 25]
```diff
@@ -53,8 +64,189 @@ tearDown()
...
+    fi
+  done
+}
+
+# To test the order of config file loading, we will put a file named
+# kworkflow.config in each place, in order, and remove the previous one.
+# The order is: PWD, XDG_CONFIG_HOME, XDG_CONFIG_DIRS, KW_ETC_DIR
+function test_parse_configuration_files_loading_order()
+{
+  build_config=()
```
Comment (rodrigosiqueira) [inline tests/build_test.sh:123]: Move this variable, after the `local` variables.

File: tests/build_test.sh [hunk 26]
```diff
@@ -53,8 +64,189 @@ tearDown()
...
+
+  local expected
+  local original_dir="$PWD"
+
+  KW_ETC_DIR='1'
+  XDG_CONFIG_DIRS='2:3:4'
+  XDG_CONFIG_HOME='5'
```
Comment (rodrigosiqueira) [inline tests/build_test.sh:130]: You probably want to restore the original values in the `setUp` function to avoid weird results in other tests.

File: tests/build_test.sh [hunk 27]
```diff
@@ -53,8 +64,189 @@ tearDown()
...
+    "3/$KWORKFLOW/$BUILD_CONFIG_FILENAME"
+    "2/$KWORKFLOW/$BUILD_CONFIG_FILENAME"
+    "5/$KWORKFLOW/$BUILD_CONFIG_FILENAME"
+  )
+
+  output="$(
+    function parse_configuration()
+    {
+      printf '%s\n' "$1"
+    }
```
Comment (rodrigosiqueira) [inline tests/build_test.sh:144]: Is this a mock function?

File: tests/build_test.sh [hunk 28]
```diff
@@ -53,8 +64,189 @@ tearDown()
...
+{
+  build_config=()
+
+  local -A shown_options
+  local -A possible_options
+  local output
+
+  # get all assigned options, including commented ones
+  # remove #'s and ='s to get option names
+  output="$(cat "$original_dir/etc/build_template.config")"
```
Comment (rodrigosiqueira) [inline tests/build_test.sh:183]: You can use this optimization:
https://kworkflow.org/content/codingstyle.html#command-substitution-and-arithmetic-expression

File: tests/build_test.sh [hunk 29]
```diff
@@ -53,8 +64,189 @@ tearDown()
...
+{
+  build_config=()
+
+  local current_path="$PWD"
+  local -a expected
+
+  function mock_parse_configuration()
+  {
+    :
+  }
```
Comment (rodrigosiqueira) [inline tests/build_test.sh:220]: Do we need this?

File: tests/build_test.sh [hunk 30]
```diff
@@ -53,8 +64,189 @@ tearDown()
...
+  for option in "${!shown_options[@]}"; do
+    if [[ ! -v possible_options["$option"] ]]; then
+      fail "($LINENO): show_variable is showing $option not present in kworkflow_template.config"
+    fi
+  done
+}
+
+function test_load_build_config()
+{
+  build_config=()
```
Comment (rodrigosiqueira) [inline tests/build_test.sh:212]: move this after the `local` variables

File: tests/kw_config_loader_test.sh [hunk 31]
```diff
@@ -16,12 +16,13 @@ function setUp()
 
 function tearDown()
 {
-  rm -rf "$TMP_DIR"
+  rm -rf "$TMP_DIR" "$SHUNIT_TMPDIR"
```
Comment (rodrigosiqueira) [inline tests/kw_config_loader_test.sh:19]: Check `$TMP_DIR` path before removing it, just in case...

File: documentation/man/features/build.rst [hunk 32]
```diff
@@ -9,6 +9,8 @@ SYNOPSIS
 | *kw* (*b* | *build*) [(-i | \--info)] [\--alert=(s | v | (sv | vs) | n)]
 | *kw* (*b* | *build*) [(-n | \--menu)] [\--alert=(s | v | (sv | vs) | n)]
 | *kw* (*b* | *build*) [(-d | \--doc)] [\--alert=(s | v | (sv | vs) | n)]
+| *kw* (*b* | *build*) [(-c | \--cpu-scaling)] <percentage> [\--alert=(s | v | (sv | vs) | n)]
```
Comment (rodrigosiqueira) [inline documentation/man/features/build.rst:12]: I think it should be:
`[(-c | \--cpu-scaling) <percentage>]`
@kwy95 ?

File: src/build.sh [hunk 33]
```diff
@@ -81,23 +83,29 @@ function kernel_build()
...
+
+  optimizations="-j$((parallel_cores * cpu_scaling_factor / 100))"
+
+  if [[ -n "${options_values['CCACHE']}" ]]; then
+    optimizations="CC=\"ccache gcc -fdiagnostics-color\" $optimizations"
   fi
 
   if [[ -n "$doc_type" ]]; then
-    command="make -j$parallel_cores $doc_type"
+    command="make $optimizations $doc_type"
```
Comment (rodrigosiqueira) [inline src/build.sh]: https://kworkflow.org/content/codingstyle.html#string-concatenation

File: src/build.sh [hunk 34]
```diff
@@ -81,23 +83,29 @@ function kernel_build()
...
     cmd_manager "$flag" "$command"
     return
   fi
 
+  command="make $optimizations ARCH=$arch $cross_compile"
```
Comment (rodrigosiqueira) [inline src/build.sh]: https://kworkflow.org/content/codingstyle.html#string-concatenation
(same for all commits)

File: src/build.sh [hunk 35]
```diff
@@ -150,6 +159,21 @@ function parse_build_options()
...
         shift
         ;;
+      --cpu-scaling | -c)
+        if [[ ! "$2" =~ [0-9]+ ]]; then
+          options_values['ERROR']="$2"
+          return 22 # EINVAL
+        fi
+        if [[ "$2" -gt 100 ]]; then
+          complain 'Upscaling CPU performance in compilation tasks may have unintended consequences!'
+        fi
```
Comment (rodrigosiqueira) [inline src/build.sh:194]: @kwy95 I think we discussed something like this in the `kw mail`. Could you review this part?

File: src/build.sh [hunk 36]
```diff
@@ -209,7 +233,9 @@ function build_help()
     '  build - Build kernel' \
     '  build (-n | --menu) - Open kernel menu config' \
     '  build (-i | --info) - Display build information' \
-    '  build (-d | --doc) - Build kernel documentation'
+    '  build (-d | --doc) - Build kernel documentation' \
+    '  build (-c | --cpu-scaling) <percentage> - Scale CPU usage by factor' \
+    '  build (--ccache) - Enable use of ccache'
```
Comment (rodrigosiqueira) [inline src/build.sh]: Just for curiosity, what will happen in these scenarios:
1. `kw b -n -c`
2. `kw b -n --ccache`
3. `kw b -d --ccache`
4. `kw b -i -c`
5. `kw b -i --ccache`

File: tests/build_test.sh [hunk 37]
```diff
@@ -340,10 +345,42 @@ function test_kernel_build_x86()
...
 }
 
+function test_kernel_isolated_build_options()
+{
+  local expected_result
+  local output
+
+  output=$(kernel_build 'TEST_MODE' --ccache | tail -n +1 | head -2)
+  declare -a expected_cmd=(
+    "make -j ARCH=x86_64  olddefconfig --silent"
```
Comment (rodrigosiqueira) [inline tests/build_test.sh]: Use single quotes.

File: tests/build_test.sh [hunk 38]
```diff
@@ -340,10 +345,42 @@ function test_kernel_build_x86()
...
+
+  output=$(kernel_build 'TEST_MODE' --cpu-scaling 50 | tail -n +1 | head -2)
+  declare -a expected_cmd=(
+    "make -j ARCH=x86_64  olddefconfig --silent"
+    "make -j$SCALING ARCH=x86_64"
+  )
+  compare_command_sequence '' "($LINENO)" "expected_cmd" "$output"
+}
+
+function test_kernel_2composed_build_options()
```
Comment (rodrigosiqueira) [inline tests/build_test.sh]: `2` is a typo?

File: tests/build_test.sh [hunk 39]
```diff
@@ -340,10 +345,42 @@ function test_kernel_build_x86()
...
+{
+  local expected_result
+  local output
+
+  output=$(kernel_build 'TEST_MODE' --ccache --cpu-scaling 50 | tail -n +1 | head -2)
+  declare -a expected_cmd=(
+    "make -j ARCH=x86_64  olddefconfig --silent"
```
Comment (rodrigosiqueira) [inline tests/build_test.sh]: ''

File: src/build.sh [hunk 40]
```diff
@@ -69,11 +70,13 @@ function kernel_build()
   fi
 
   if [[ -n "$cross_compile" ]]; then
-    cross_compile="CROSS_COMPILE=$cross_compile"
+    cross_compile=" CROSS_COMPILE=$cross_compile "
```
Comment (rodrigosiqueira) [inline src/build.sh:78]: https://kworkflow.org/content/codingstyle.html#string-concatenation

File: documentation/man/features/build.rst [hunk 41]
```diff
@@ -51,6 +54,18 @@ OPTIONS
   This option allows the user to enable ``ccache`` usage during compilation
   tasks, which should improve compile times in subsequent compilations. 
 
+-w, \--warnings=(1 | 2 | 3 | 12 | 13 | 23 | 123):
+  This can be used to enable compilation warnings accordingly. Please check the
+  kernel's ``make help`` for more info.
```
Comment (rodrigosiqueira) [inline documentation/man/features/build.rst:59]: I think it will be more friendly if you explain the warning level 1 and 2... Actually, 1 and 2 are hard to understand; maybe we could have a string alias that is easy to understand.

File: documentation/man/features/build.rst [hunk 42]
```diff
@@ -51,6 +54,18 @@ OPTIONS
...
+
+-s, \--save-log-to=path:
```
Comment (rodrigosiqueira) [inline documentation/man/features/build.rst:61]: Do we need `=`? I think it would be better if we didn't need it.

File: documentation/man/features/build.rst [hunk 43]
```diff
@@ -51,6 +54,18 @@ OPTIONS
...
+  This option will save the full compilation log with the enabled warnings to
+  the specified path.
```
Comment (rodrigosiqueira) [inline documentation/man/features/build.rst:63]: Can we create a default file if the user does not provide a path? Also, we should document this behavior here.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hi @isinyaaa ,
Thanks a lot for your PR; that's an awesome change!

File: src/build.sh [hunk 44]
```diff
@@ -166,6 +166,38 @@ function parse_build_options()
...
 
+# This function loads the kw build configuration files into memory, populating
+# the $build_config hashtable. The files are parsed in a specific order,
+# allowing higher level setting definitions to overwrite lower level ones.
+function load_build_config()
+{
+  local -a config_dirs
+  local config_dirs_size
+
+  IFS=: read -ra config_dirs <<< "${XDG_CONFIG_DIRS:-"/etc/xdg"}"
```
[the remark this answers is not in the document]
Comment (isinyaaa) [reply src/build.sh:248]: from what I understand, it should split the `XDG_CONFIG_DIRS` string and save each entry separately in the `config_dirs` array. Then, if there's no such variable, use the `/etc/xdg` as default.

As you noted, I feel like this code could improve, but the original one is like this as well, so I just copy pasted it, actually.

File: tests/build_test.sh (hunk 24 above)
Comment (isinyaaa) [reply tests/build_test.sh:103]: Actually, this is copy pasted, so I'm not sure about much of its workings, but I can try to improve it.

File: tests/build_test.sh (hunk 23 above)
Comment (isinyaaa) [reply tests/build_test.sh:99]: I think it might actually work that way because `lineno=LINENO` but then you'd have to replace it twice when you use it, like `${!lineno}` (like a double pointer).

File: tests/build_test.sh (hunk 24 above)
Comment (isinyaaa) [reply tests/build_test.sh:103]: I've done a quick research on this code, and could understand the following:
1. It was introduced by @matheustavares in 8e716c50, and
2. It aims at generalizing the process you use to verify those config arrays, the basic idea being that:
  1. You access array B using a key `$k` that was taken from array A.
  2. If B has such a key, then you end up with the string "token", otherwise it's empty (which is obviously different from "token" on the right hand side).
  3. Now you can check every key in the array easily.

Truly an ingenious solution!

File: tests/build_test.sh (hunk 27 above)
Comment (isinyaaa) [reply tests/build_test.sh:144]: seems like it

**Comment by rodrigosiqueira** [comment]
Hi @isinyaaa , could you update this PR?

**Comment by isinyaaa** [comment]
> Hi @isinyaaa , could you update this PR?

I still have many comments to address, but this is a partial update to get these commits up here.

File: tests/plugins/kernel_install/utils_test.sh [hunk 45]
```diff
@@ -72,6 +72,7 @@ function setUp()
 function tearDown()
 {
   rm -rf "$SHUNIT_TMPDIR"
+  mkdir -p "$SHUNIT_TMPDIR"
```
Comment (rodrigosiqueira) [inline tests/plugins/kernel_install/utils_test.sh:75]: I think it makes more sense to move this `mkdir -p` from `terarDown` to `setUp`.

File: src/kw_config_loader.sh [hunk 46]
```diff
@@ -158,11 +158,18 @@ function parse_configuration()
 # higher level setting definitions to overwrite lower level ones.
 function load_configuration()
 {
+  if [[ -v configurations && "$OVERRIDE_CONFIG" != 1 ]]; then
+    return
+  fi
+
```
Comment (rodrigosiqueira) [inline src/kw_config_loader.sh:162]: The commit message here does not describe anything... Please, update the commit message.

File: etc/build_template.config [hunk 47]
```diff
@@ -0,0 +1,17 @@
+# Specify the default architecture used by kw
```
Comment (rodrigosiqueira) [inline etc/build_template.config:1]: I recently introduced the `kw init --template` and this commit will be affected by this feature. You will also need to update the template feature to copy the `build.config` file correctly.

1. https://kworkflow.org/man/features/init.html

File: src/build.sh [hunk 48]
```diff
@@ -114,6 +114,50 @@ function kernel_build()
...
+# allowing higher level setting definitions to overwrite lower level ones.
+function load_build_config()
+{
+  if [[ -v build_config && "$OVERRIDE_BUILD_CONFIG" != 1 ]]; then
+    unset OVERRIDE_BUILD_CONFIG
+    return
+  fi
+
+  local -a config_dirs
+  local config_dirs_size
```
Comment (rodrigosiqueira) [inline src/build.sh:158]: Let's keep variables at the beginning of the function.

File: src/build.sh [hunk 49]
```diff
@@ -114,6 +114,50 @@ function kernel_build()
...
+
+  if [[ -v XDG_CONFIG_DIRS ]]; then
+    IFS=: read -ra config_dirs <<< "$XDG_CONFIG_DIRS"
+  else
+    [[ -d '/etc/xdg' ]] && config_dirs=('/etc/xdg')
+  fi
+
+  # Old users may not have split their configs yet
```
Comment (rodrigosiqueira) [inline src/build.sh:166]: Is this comment related to the below line? It looks unrelated to me.

File: src/build.sh [hunk 50]
```diff
@@ -114,6 +114,50 @@ function kernel_build()
...
+
+  parse_configuration "${XDG_CONFIG_HOME:-"$HOME/.config"}/$KWORKFLOW/$BUILD_CONFIG_FILENAME" build_config
+
+  if [[ -f "$PWD/$KW_DIR/$BUILD_CONFIG_FILENAME" ]]; then
+    parse_configuration "$PWD/$KW_DIR/$BUILD_CONFIG_FILENAME" build_config
+  else
+    # Old users may not have used kw init yet, so they wouldn't have .kw
+    warning "Please use kw init to update your config files"
+  fi
+}
```
Comment (rodrigosiqueira) [inline src/build.sh:189]: Can we make this function something generic? I guess other features will benefit from that (e.g., deploy).

File: src/kw_config_loader.sh [hunk 51]
```diff
@@ -1,7 +1,9 @@
 include "$KW_LIB_DIR/kwio.sh"
 include "$KW_LIB_DIR/kwlib.sh"
 
+# these should be the default names for config files
 CONFIG_FILENAME=kworkflow.config
+BUILD_CONFIG_FILENAME=build.config
```
Comment (rodrigosiqueira) [inline src/kw_config_loader.sh:6]: Add single quotes around `build.config`

File: documentation/man/features/build.rst (hunk 32 above)
Comment (rodrigosiqueira) [reply documentation/man/features/build.rst:12]: @kwy95

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hi @isinyaaa ,
This PR is a really welcome one! I added some other comments, but let's try to complete it :)

**Comment by rodrigosiqueira** [comment]
@isinyaaa any update here?

**Comment by isinyaaa** [comment]
I just needed to rebase this on my current system, so this is not done yet.

File: documentation/man/features/build.rst (hunk 32 above)
Comment (isinyaaa) [reply documentation/man/features/build.rst:12]: indeed you're right

File: src/build.sh (hunk 35 above)
Comment (isinyaaa) [reply src/build.sh:240]: @kwy95 ping

File: src/build.sh (hunk 36 above)
Comment (isinyaaa) [reply src/build.sh]: > 1. `kw b -n -c`
> 2. `kw b -n --ccache`

`-n` will not take any of these flags into consideration.

> 3. `kw b -d --ccache`

This should take `ccache` into consideration just fine :)

> 4. `kw b -i -c`
> 5. `kw b -i --ccache`

I'm really not sure about the `-i` flag, as I've not messed with it at all.

File: tests/build_test.sh (hunk 38 above)
Comment (isinyaaa) [reply tests/build_test.sh]: nope

File: documentation/man/features/build.rst (hunk 42 above)
Comment (isinyaaa) [reply documentation/man/features/build.rst:61]: we actually don't use `=`, this was a typo :)

File: src/build.sh [hunk 52]
```diff
@@ -179,3 +211,45 @@ function build_help()
     '  build (-i | --info) - Display build information' \
     '  build (-d | --doc) - Build kernel documentation'
 }
+
+# This function is used to show the current set up used by kworkflow.
+function show_build_variables()
+{
```
[the remark this answers is not in the document]
Comment (isinyaaa) [reply src/build.sh:308]: I tried my hand at it but still haven't tested it. I have one question though: when should those variables be printed?

File: documentation/man/features/build.rst (hunk 41 above)
Comment (isinyaaa) [reply documentation/man/features/build.rst:59]: Agree!

**Comment by isinyaaa** [comment]
@rodrigosiqueira I'll update this PR with the comments you provided, but I think it'd be best if we somehow split it to make revision easier, as it's already given us so much to talk about.

File: documentation/man/features/build.rst (hunk 32 above)
Comment (kwy95) [reply documentation/man/features/build.rst:12]: My bad, I missed this comment somehow.

File: src/build.sh (hunk 35 above)
Comment (kwy95) [reply src/build.sh:240]: Hi, the relevant function in the mail.sh file is `reposition_commit_count_arg` and the end of the parser as well, if you want take a look at it. I don't think the use case is the same, but it might help somehow. Anyway, I will review this section further in the new PR.
